ceph: arrow 19 compilation fix with protobuf 30; snappy: restore of RTTI patch

benaryorg opened the pull request:
It seems that Arrow 20 which broke compilation initially causes issues with our tests when the patch which was merged upstream is applied. At the same time the bundled Arrow 17 can't work with current Re2 due to its C++ version. The thing that broke the Arrow 19 build is because Arrow 19 is itself not compatible with Protobuf 30, which is pulled in somewhere along the dependency graph by something else causing linking issues. However, the patch for Protobuf 30 compat applies cleanly to our Arrow 19, which means we can dump the older version pin and instead use this patch to get Arrow 19 working again.

---

Resolves #426401
Supersedes #426609 (here's to hoping the next Ceph major version will have less issues with arrow)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

benaryorg committed:
ceph: arrow 19 compilation fix with protobuf 30

It seems that Arrow 20 which broke compilation initially causes issues with our tests when the patch which was merged upstream is applied.
At the same time the bundled Arrow 17 can't work with current Re2 due to its C++ version.
The thing that broke the Arrow 19 build is because Arrow 19 is itself not compatible with Protobuf 30, which is pulled in somewhere along the dependency graph by something else causing linking issues.
However, the patch for Protobuf 30 compat applies cleanly to our Arrow 19, which means we can dump the older version pin and instead use this patch to get Arrow 19 working again.

Signed-off-by: benaryorg <binary@benary.org>

nh2 commented on pkgs/tools/filesystems/ceph/arrow-cpp-19.nix:
```diff
@@ -101,6 +99,15 @@ stdenv.mkDerivation (finalAttrs: {
+  patches = [
+    (fetchpatch2 {
+      name = "protobuf-30-compat.patch";
+      url = "https://github.com/apache/arrow/pull/46136.patch";
```
Probably better to set this to `https://github.com/apache/arrow/compare/a3a0c16a823387537243d513ba89d853d9b5eed6~..de11a52b9f57c89f02fabcf65ec80d0fe4d54562.patch` than the PR number. Seems to have the same `hash`, so should not change the eval.

benaryorg replied:
Just checked the *CONTRIBUTING.md* and there doesn't seem to be any specific guidance on that other than not referencing *open* PRs (for obvious reasons).
Personally I prefer linking to the actual PR so it's easier to figure out where stuff came from, and since the PR is merged it can't change or be reopened AFAIK, but I'm fully okay changing this.

I'll have to clean up this PR before undrafting anyway, so I'll get back to that later (just pushed the revert in here to have a canonical branch to test against).

nh2 replied:
@benaryorg I also think either is OK, it really is a nitpick; the nice thing about the commit based URL is that it allows to figure out what's in it even if the upstream does weird things such as reopening the PR and force-pushing the branch (not sure why somebody would want to do that htough), or deleting the whole repo (thus making it slightly less dependentent on info stored only on Github).

benaryorg committed:
snappy: restore of RTTI patch

This is required for *Ceph* to compile, as the comment states, and was accidentally removed.
A similar patch is also present for *leveldb* for instance.
Also bumps the patch to the appropriate version.

partial revert of 5240499fb672c5f7e25cd84bff780062cb3cf75a

Signed-off-by: benaryorg <binary@benary.org>

benaryorg commented:
- changed the target to *staging* due to the number of rebuilds
- as discussed in the associated issue the snappy patch was reintroduced without a downgrade
- bump said patch to the current sidestream patch version
- I'll mark this as ready for review (and update the OP with the currently commented out "Resolves" and "Supersedes") statements after [the Hydra tests](https://hydra.cloud.bsocat.net/eval/432515) have succeeded

nh2 commented:
@trofi @pbsds in this PR we revert the removal of the `snappy` RTTI patch from:

https://github.com/NixOS/nixpkgs/pull/406663#pullrequestreview-3056626787

Please point out if for some reason there might a problem with that (even better, do you know why that was removed in the first place?).

trofi commented:
> @trofi @pbsds in this PR we revert the removal of the `snappy` RTTI patch from:
> 
> [#406663 (review)](https://github.com/NixOS/nixpkgs/pull/406663#pullrequestreview-3056626787)
> 
> Please point out if for some reason there might a problem with that (even better, do you know why that was removed in the first place?).

To copy https://github.com/NixOS/nixpkgs/pull/406663#discussion_r2233698891 here:

I think (I don't remember' sorry) I removed the `RTTI` patch because it was rejected upstream https://github.com/google/snappy/pull/144  and was failing to apply downstream.

benaryorg commented:
> I think (I don't remember' sorry) I removed the `RTTI` patch because it was rejected upstream and was failing to apply downstream.

I'll add a comment for that then to avoid it happening in the future.
The referenced patch is revisioned and regularly updated but that's not readily apparent, but usually bumping it takes a bit of clicking around to get the right rev off the site and the new one should apply.

benaryorg commented:
@trofi seems I can't request a review since you're not a member of the GitHub group, so I'll ping you directly; could you please just give a nod on whether or not the comment in this version is sufficient information to convey the importance of the patch, and also to provide clear enough instructions to keep it updated in the future (the site feels a bit tedious to navigate on if you're looking for the current rev IMHO)?

@nh2 if you could have a glance at the PR in its current state, and double check the Ceph side of things that'd be great.

If both of you approve the PR I think we can merge it given that it targets *staging* anyway, so we'll see any failures on upstream Hydra, and we can fix those using commits to *staging-next* to avoid extra cycles and get that merged into *master* soon (enough).

trofi commented:
> @trofi seems I can't request a review since you're not a member of the GitHub group, so I'll ping you directly; could you please just give a nod on whether or not the comment in this version is sufficient information to convey the importance of the patch, and also to provide clear enough instructions to keep it updated in the future (the site feels a bit tedious to navigate on if you're looking for the current rev IMHO)?

Restoring the patch sounds fine to me.

But to address longer term maintainability health it would be nice if `ceph` would not require patched source of `snappy`.

benaryorg commented:
> But to address longer term maintainability health it would be nice if `ceph` would not require patched source of `snappy`.

As far *as I understand* (which may well be wrong), any application using *dlopen* but still requiring to extend a certain class does require that patch, and upstream has rejected that patch on grounds that they want to use the same compilation environment as Google Chrome regardless of breaking other software.
My impression is that Ceph's usage is legitimate and absolutely within what software normally does (given that Ceph uses a lot of Python the *dlopen* part does make sense).

<details><summary>grumble</summary>

I do not see a future where the patch is not needed, and also given that [Google is a very small startup](https://www.nasdaq.com/market-activity/stocks/goog) with limited resources and [barely any involvement with build systems or CI](https://en.wikipedia.org/wiki/Bazel_(software)) I fully understand why they simply cannot afford the maintenance burden of a change which removes code actively messing with provided compile options.
If upstream [provided those options at compile time externally](https://github.com/NixOS/nixpkgs/blob/63ba1c64e697b6dc48754ea62aeaa8c33d04c0f5/pkgs/by-name/c2/c2ffi/package.nix#L52) (i.e. within their own build) instead of literally search and replacing them within cmake, none of this would be required in the first place but – again – such [a small company sometimes has to cut corners](https://www.trueup.io/co/google#layoffs) lest they [lose out to bigger competitors](https://searchengineland.com/google-search-market-share-drops-2024-450497) on the market.

</details>

trofi commented:
Added the comment in https://github.com/google/snappy/issues/189#issuecomment-3124551789

nh2 commented:
I also think it's not a big thing to carry that CMake-only patch perpetually: Red Hat can have their dislike for static linking and Google can have a dislike for `dlopen()`; adjusting that on our side is OK in my opinion, if the changes stay indeed a 2-line `CMakeLists.txt` diff.

@benaryorg I think the added comment

```
    # While the patch was rejected upstream, it does not make it any less necessary to carry forward.
    # ==> lack of RTTI *breaks* Ceph (and others) <==
```

is very good and makes the story clear.

nh2 reviewed (approved):
This looks good to me. Let's ship it to `staging`.

nh2 commented:
The ofborg failed with `error: Nix daemon disconnected unexpectedly (maybe it crashed?)`

nh2 reviewed (changes requested):
Setting back to "request changes" just to check where the hash mismatch comes from; beyond that it's still good to merge.

benaryorg commented:
I just tested the branch rebased onto an older commit of *master* (1a689bc79a3f566b8f7e8384ffcdbf2b05026314) and it built, however given the merge to staging this wants to rebuild everything starting at *curl* which is a whole damn lot.
